### ISSUE TITLE
400 Bad Request Not Being Caught

### ISSUE THREAD

**Issue by webtechy** [SharePoint/PnP-JS-Core#375]
#### Category
[ ] Enhancement
[x] Bug
[x] Question

#### Expected / Desired Behavior / Question
I would expect using catch() and onRejected() methods that the error would be caught.

#### Observed Behavior
No pop-up error to the use, just an console.error message:

> /_api/web/lists/getByTitle('Requirements')/items(65) 400 (Bad Request)

#### Steps to Reproduce

```
    public updateRecord() {
        window.console && console.log('updateRecord(): Starting ...');

        var self = this;

        if (this.RegulatoryProvisionsPicked.length > 1)
        {
            window.console && console.warn("updateRecord(): More than 1 regulatory provision picked.");
            self.pageError = true;
            self.pageErrorMessage = "You can only choose a single regulatory provision.";
        }
        else
        {
            pnp.sp.web.lists.getByTitle("Requirements").items.getById(this.Requirement.Id).update({
                Title: self.Requirement.Title,
                Requirement_x0020_Details: self.Requirement.Details,
                //R2_x0020_Category: this.Requirement.Category,
                Requirement_x0020_RatingValue: self.Requirement.Rating.RatingText,
                Final_x0020_Version: self.Requirement.FinalVersion,
                Regulatory_x0020_ProvisionId: self.Requirement.RegulatoryProvision.Id
            }).then((reqResult) => {
                window.console && console.log("updateRecord(): reqResult ...");
                window.console && console.log(reqResult);
                // Save attachments to this item.
                window.console && console.log("updateRecord(): this.Requirement.Attachments ...");
                window.console && console.log(self.Requirement.Attachments);
                if (self.NewAttachments != null)
                {
                    let item = pnp.sp.web.lists.getByTitle("Requirements").items.getById(self.Requirement.Id);
                    self.NewAttachments.forEach(attachment => {
                        item.attachmentFiles.add(attachment.name, attachment.rawFile).then(attachResult => {
                            window.console && console.log("updateRecord(): Attaching file (" + attachment.name + ")...");
                            window.console && console.log(attachResult);
                            self.Requirement.Attachments.push({
                                name: attachment.name,
                                size: attachment.size
                            })
                        },
                        function(rejected) {
                            window.console && console.error("updateRecord(): E1 - " + rejected.message);
                            self.pageError = true;
                            self.pageErrorMessage = "updateRecord(): E1 - " + rejected.message;
                            window.console && console.log("updateRecord(): Remove temporary uploads ...");
                        }).catch((e) => {
                            window.console && console.error("updateRecord(): E2 - " + e.message); 
                            self.pageError = true;
                            self.pageErrorMessage = "updateRecord(): E2 - " + e.message;
                        }); 
                    });
                    // OK, all uploaded, let's clear the temporary uploads.
                    self.NewAttachments = [];
                }

                window.console && console.log('updateRecord(): Requirement saved (updated).');
                self.Requirement.Json = JSON.stringify(reqResult.data);
            },
            function(rejected) {
                window.console && console.error("updateRecord(): E3 - " + rejected.message);
                self.pageError = true;
                self.pageErrorMessage = "updateRecord(): E3 - " + rejected.message;  
            }).catch((e) => {
                window.console && console.error("updateRecord(): E4 - " + e.message);
                self.pageError = true;
                self.pageErrorMessage = "updateRecord(): E4 - " + e.message;
            });
            self.pageMode = "Display";
        }
    }
```

**Comment by patrick-rodgers** [comment]
Is there a reason you are using both, would simplify things to do either a catch or the onrejected delegate? I usually just use catch but YMMV. You might be swallowing the error above in the if block. I am guessing the `item.attachmentFiles.add` is failing? If so try removing the handlers from it or throw an error in there just to see if it hits the outer handlers.

**Comment by webtechy** [comment]
I assumed there would be instances where a call would be rejected but not necessarily be caught in a catch or vice versa?

**Comment by patrick-rodgers** [comment]
The .catch and the onrejected handler both do the same thing. I see more folks using .catch, I think it is a little cleaner and you can have a single catch at the end of a long promise chain and it will get all the errors anywhere in the chain. That doesn't make that way "right" just a preference. Did you try throwing an exception in the inner body?

**Comment by elebetsamer** [comment]
I'm doing something pretty simple and also getting an uncaught `400: Bad Request`.

In the example below, if I `updateData` contains valid fields from the list, the request succeeds. But if I use the name of a field that doesn't exist, I get a regular old javascript error and don't get in to my catch block.

```javascript
var updateData = { test: 'testing' };

$pnp.sp.web.lists.getByTitle('Pages').items.getById(_spPageContextInfo.pageItemId).update(updateData).then(function(result) {
  console.log('Success');
}).catch(function(response) {
  console.error('Something went wrong');
});```

**Comment by elebetsamer** [comment]
Nevermind, I just updated to **2.0.2** and that seems to have fixed it. This seems to have been fixed in #329

**Comment by webtechy** [comment]
I have come to the conclusion that 4xx responses do not get caught (only 5xx for example). I am currently running `"sp-pnp-js": "^2.0.1"` so shall also try upgrading.

**Comment by patrick-rodgers** [comment]
There was a bug as @elebetsamer linked specific to item.update where we were eating the error (bad) and that is fixed in 2.0.2. We should not be eating any errors in the library and all exceptions should be catchable in your code. I think in this case if you upgrade to 2.0.2 you should be all set in this case. Going to close this as resolved, please reopen if you find that to not be the case. Thanks!
